kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -66,245 +82,410 @@ String filterDescriptor() {
     }
 
     @Override
-    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
-        if (msg instanceof InternalRequestFrame<?>) {
-            // jump the queue, internal request must flow!
-            doWrite(ctx, msg, promise);
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        this.ctx = ctx;
+        super.channelActive(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) {
+        if (msg instanceof InternalResponseFrame<?> decodedFrame) {
+            // jump the queue, let responses to asynchronous requests flow back to their sender
+            if (decodedFrame.isRecipient(filter)) {
+                completeInternalResponse(decodedFrame);
+            }
+            else {
+                readDecodedResponse(decodedFrame);
+            }
         }
-        else if (writeFuture.isDone()) {
-            writeFuture = doWrite(ctx, msg, promise);
+        else if (msg instanceof DecodedResponseFrame<?> decodedFrame) {
+            if (readFuture.isDone()) {
+                readFuture = readDecodedResponse(decodedFrame);
+            }
+            else {
+                readFuture = readFuture.whenComplete((a, b) -> {
+                    if (ctx.channel().isOpen()) {
+                        readDecodedResponse(decodedFrame);
+                    }
+                });
+            }
         }
         else {
-            writeFuture = writeFuture.whenComplete((a, b) -> {
-                if (ctx.channel().isOpen()) {
-                    doWrite(ctx, msg, promise);
-                }
-            });
+            if (!(msg instanceof OpaqueResponseFrame)) {
+                throw new IllegalStateException("Unexpected message reading from upstream:  " + msg);
+            }
+            ctx.fireChannelRead(msg);
```

Comment:
Opaque frames need to go on the queue too, right? If this filter is awaiting some async computation triggered by a decoded frame we don't want an OpaqueResponseFrame to be forwarded on to the client and responding out of order.

Reply:
PR here https://github.com/k-wall/kroxylicious/pull/6

Reply:
Great catch. I've added to this PR.